packages/openapi3/src/openapi.ts

```diff
@@ -1021,15 +1018,12 @@ function createOAPIEmitter(program: Program, options: ResolvedOpenAPI3EmitterOpt
 
     const discriminator = getDiscriminator(program, model);
     if (discriminator) {
-      if (!validateDiscriminator(discriminator, derivedModels)) {
-        // appropriate diagnostic is generated with the validate function
-        return {};
-      }
+      const [union, diagnostics] = getDiscriminatedUnion(model, discriminator);
+      program.reportDiagnostics(diagnostics);
```

Comment:
Feel like this should already be one with an `$onValidate` for all discriminated types. Though this would be the first time we have an $onValidate in the compiler. Any opinions here?

Reply:
onValidate feels very late to be throwing this error, but I suppose the issue is that for the model form, we don't have a good time to check for this since a new model could show up at any time that extends a discriminator? If we had some kind of "run this validation whenever a model extends me" hook, that could work, but without that we just have to check at the very end?

Given this is a built-in feature of the compiler, maybe just adding an explicit pass in code to check for this, before onValidate is even called, might be ok as well.

Reply:
yeah that's the problem don't have the full picture needed until everything has been checked. I'll update to have an special check for this after running the checker. 

Not exactly the same as you are suggesting but I was thinking that it would be also nice to have a way for decorators to say run this validation after all the decorators have run.

```ts

export function $myDec() { 
 // do stuff

//return some post validation
return () => {
  // here I can make sure some other decorator wasn't called.
}
}
```

Maybe that pattern could be expanded to more scenarios but this also feels like its getting very complex

Reply:
I think it's reasonable to make this an explicit pass before the onValidate is called. I suppose we could also ensure our $onValidate handlers run first somehow? I don't have a strong opinion here.